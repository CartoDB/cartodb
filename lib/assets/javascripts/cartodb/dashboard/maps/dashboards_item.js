var cdb = require('cartodb.js');
var moment = require('moment');
var navigateThroughRouter = require('../../common/view_helpers/navigate_through_router');
var Utils = require('cdb.Utils');
cdb.admin = require('cdb.admin');

/**
 * Represents a map card on dashboard.
 */
module.exports = cdb.core.View.extend({

  className: 'MapsList-item',
  tagName: 'li',

  initialize: function() {
    this.router = this.options.router;
    this.user = this.options.user;
    this.template = cdb.templates.getTemplate('dashboard/views/dashboards_item');
  },

  render: function() {
    this.clearSubViews();
    var url = this.model.dashboardURL();

    this.$el.html(
      this.template({
        url: url,
        name: this.model.get('name'),
        description: this.model.get('description'),
        timeDiff: moment(this.model.get('updated_at')).fromNow(),
      })
    );

    return this;
  }

});
