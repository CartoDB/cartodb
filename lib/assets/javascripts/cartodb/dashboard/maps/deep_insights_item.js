var cdb = require('cartodb.js-v3');
var moment = require('moment-v3');
var navigateThroughRouter = require('../../common/view_helpers/navigate_through_router');
var Utils = require('cdb.Utils');
cdb.admin = require('cdb.admin');

/**
 * Represents a map card on dashboard.
 */
module.exports = cdb.core.View.extend({

  className: 'MapsList-item',
  tagName: 'li',

  initialize: function() {
    this.router = this.options.router;
    this.user = this.options.user;
    this.template = cdb.templates.getTemplate('dashboard/views/deep_insights_item');
  },

  render: function() {
    this.clearSubViews();
    var url = this.model.deepInsightsUrl(this.user);

    this.$el.html(
      this.template({
        url: url,
        title: this.model.get('title'),
        description: this.model.get('description'),
        timeDiff: moment(this.model.get('updated_at')).fromNow(),
      })
    );

    return this;
  }

});
