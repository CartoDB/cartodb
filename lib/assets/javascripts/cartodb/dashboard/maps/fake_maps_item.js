var cdb = require('cartodb.js-v3');
cdb.admin = require('cdb.admin');

/**
 *  Fake map card on dashboard
 */

module.exports = cdb.core.View.extend({

  className: 'MapsList-item',
  tagName: 'li',

  initialize: function() {
    this.template = cdb.templates.getTemplate('dashboard/views/fake_maps_item');
  },

  render: function() {
    this.$el.html(this.template());
    return this;
  }

});
