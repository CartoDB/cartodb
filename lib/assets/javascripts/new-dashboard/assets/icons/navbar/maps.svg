<svg xmlns="http://www.w3.org/2000/svg" width="19" height="21">
  <path d="M18.555 15.832l-6.006 3.992c-.154.103-.326.176-.549.176-.223 0-.444-.074-.624-.219l-4.429-3.543-5.392 3.594C1.387 19.944 1.162 20 1 20c-.162 0-.324-.038-.472-.118C.203 19.708 0 19.369 0 19V5c0-.333.167-.646.445-.832l6-4C6.6.066 6.82 0 7.035.008c.215.008.44.094.59.211l4.429 3.544L17.446.168c.306-.205.701-.222 1.026-.049.325.173.528.512.528.881v14c0 .334-.167.647-.445.832zM2 5.535v11.597l4-2.667V2.869zm9 11.385V5.481L8 3.082V14.52zm6-2.455V2.869l-4 2.666v11.597z" fill="#fff"/>
  <path class="navbar-iconFill" d="M12 17.5V4.274L6 1.5v13.225z" fill="transparent"/>
</svg>