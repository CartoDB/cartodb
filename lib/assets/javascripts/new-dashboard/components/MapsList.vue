<template>
  <div class="container grid">
    <div class="full-width">
      <SectionTitle class="grid-cell" :showActionButton="!selectedMaps.length" ref="headerContainer">
        <template slot="icon">
          <img src="../assets/icons/section-title/map.svg">
        </template>

        <template slot="title">
          <VisualizationsTitle
            :defaultTitle="$t(`MapsPage.header.title['${appliedFilter}']`)"
            :selectedItems="selectedMaps.length"/>
        </template>

        <template slot="warning">
          <NotificationBadge type="warning" v-if="shouldShowLimitsWarning">
            <div class="warning">
              <span v-if="isOutOfPublicMapsQuota && !isOutOfPrivateMapsQuota" class="is-bold" v-html="$t('MapsPage.header.warning.counter', { counter: `${publicMapsCount}/${publicMapsQuota}`, type: `public` })"></span>
              <span v-if="isOutOfPrivateMapsQuota && !isOutOfPublicMapsQuota" class="is-bold" v-html="$t('MapsPage.header.warning.counter', { counter: `${privateMapsCount}/${privateMapsQuota}`, type: `private` })"></span>
              <span v-if="isOutOfPublicMapsQuota && isOutOfPrivateMapsQuota" class="is-bold" v-html="$t('MapsPage.header.warning.doubleCounter', { publicCounter: `${publicMapsCount}/${publicMapsQuota}`, privateCounter: `${privateMapsCount}/${privateMapsQuota}`})"></span>
              <span v-html="$t('MapsPage.header.warning.upgrade', { path: upgradeUrl })"></span>
            </div>
          </NotificationBadge>
        </template>

        <template slot="dropdownButton">
          <MapBulkActions
            :selectedMaps="selectedMaps"
            :areAllMapsSelected="areAllMapsSelected"
            v-if="hasBulkActions && selectedMaps.length"
            @selectAll="selectAll"
            @deselectAll="deselectAll"></MapBulkActions>

          <SettingsDropdown
            section="maps"
            v-if="!selectedMaps.length"
            :filter="appliedFilter"
            :order="appliedOrder"
            :orderDirection="appliedOrderDirection"
            :metadata="mapsMetadata"
            @filterChanged="applyFilter">
            <img svg-inline src="../assets/icons/common/filter.svg">
          </SettingsDropdown>

          <div class="mapcard-view-mode" @click="toggleViewMode" v-if="shouldShowViewSwitcher">
            <img svg-inline src="../assets/icons/common/compactMap.svg" v-if="!isCondensed">
            <img svg-inline src="../assets/icons/common/standardMap.svg" v-if="isCondensed">
          </div>
        </template>
        <template slot="actionButton" v-if="!isFirstTimeViewingDashboard && !selectedMaps.length">
          <CreateButton visualizationType="maps" :disabled="!canCreateMaps">{{ $t(`MapsPage.createMap`) }}</CreateButton>
        </template>
      </SectionTitle>

      <div class="grid-cell" v-if="initialState">
        <CreateMapCard></CreateMapCard>
      </div>

      <slot name="navigation"></slot>

      <div
          v-if="shouldShowListHeader"
          class="grid-cell grid-cell--noMargin grid-cell--col12 grid__head--sticky"
          :class="{ 'has-user-notification': isNotificationVisible }">
        <CondensedMapHeader
          :order="appliedOrder"
          :orderDirection="appliedOrderDirection"
          @orderChanged="applyOrder"
          v-if="shouldShowListHeader">
        </CondensedMapHeader>
      </div>

      <ul :class="[isCondensed ? 'grid grid-cell' : 'grid']" v-if="isFetchingMaps">
        <li :class="[isCondensed ? condensedCSSClasses : cardCSSClasses]" v-for="n in maxVisibleMaps" :key="n">
          <MapCardFake :condensed="isCondensed"></MapCardFake>
        </li>
      </ul>

      <ul :class="[isCondensed ? 'grid grid-column grid-cell' : 'grid']" v-if="!isFetchingMaps && currentEntriesCount > 0">
        <li v-for="map in maps" :class="[isCondensed ? condensedCSSClasses : cardCSSClasses]" :key="map.id">
          <MapCard
            :condensed="isCondensed"
            :visualization="map"
            :isSelected="isMapSelected(map)"
            :selectMode="isSomeMapSelected"
            :canHover="canHoverCard"
            @toggleSelection="toggleSelected"
            @contentChanged="onContentChanged">
          </MapCard>
        </li>
      </ul>

      <EmptyState
        :text="emptyStateText"
        v-if="emptyState">
        <img svg-inline src="../assets/icons/common/compass.svg">
      </EmptyState>

    </div>
  </div>
</template>

<script>
import { mapState, mapGetters } from 'vuex';
import CreateButton from 'new-dashboard/components/CreateButton.vue';
import CreateMapCard from 'new-dashboard/components/CreateMapCard';
import EmptyState from 'new-dashboard/components/States/EmptyState';
import InitialState from 'new-dashboard/components/States/InitialState';
import MapBulkActions from 'new-dashboard/components/BulkActions/MapBulkActions.vue';
import MapCard from 'new-dashboard/components/MapCard/MapCard.vue';
import CondensedMapHeader from 'new-dashboard/components/MapCard/CondensedMapHeader.vue';
import MapCardFake from 'new-dashboard/components/MapCard/fakes/MapCardFake';
import SectionTitle from 'new-dashboard/components/SectionTitle';
import VisualizationsTitle from 'new-dashboard/components/VisualizationsTitle';
import NotificationBadge from 'new-dashboard/components/NotificationBadge';
import SettingsDropdown from 'new-dashboard/components/Settings/Settings';
import { shiftClick } from 'new-dashboard/utils/shift-click.service.js';

export default {
  name: 'MapsList',
  props: {
    maxVisibleMaps: {
      type: Number,
      default: 12
    },
    hasBulkActions: {
      type: Boolean,
      default: false
    },
    canHoverCard: {
      type: Boolean,
      default: false
    },
    canChangeViewMode: {
      type: Boolean,
      default: false
    },
    isCondensedDefault: {
      type: Boolean,
      default: true
    }
  },
  components: {
    CreateButton,
    CreateMapCard,
    EmptyState,
    SettingsDropdown,
    MapBulkActions,
    MapCard,
    CondensedMapHeader,
    MapCardFake,
    SectionTitle,
    VisualizationsTitle,
    NotificationBadge,
    InitialState
  },
  data () {
    return {
      selectedMaps: [],
      cardCSSClasses: 'grid-cell grid-cell--col4 grid-cell--col6--tablet grid-cell--col12--mobile map-element',
      condensedCSSClasses: 'card-condensed',
      isCondensed: this.isCondensedDefault,
      lastCheckedItem: null
    };
  },
  computed: {
    ...mapState({
      appliedFilter: state => state.maps.filterType,
      appliedOrder: state => state.maps.order,
      appliedOrderDirection: state => state.maps.orderDirection,
      maps: state => state.maps.list,
      mapsMetadata: state => state.maps.metadata,
      isFetchingMaps: state => state.maps.isFetching,
      currentEntriesCount: state => state.maps.metadata.total_entries,
      filterType: state => state.maps.filterType,
      totalUserEntries: state => state.maps.metadata.total_user_entries,
      totalShared: state => state.maps.metadata.total_shared,
      isFirstTimeViewingDashboard: state => state.config.isFirstTimeViewingDashboard,
      upgradeUrl: state => state.config.upgrade_url
    }),
    ...mapGetters({
      publicMapsQuota: 'user/publicMapsQuota',
      publicMapsCount: 'user/publicMapsCount',
      isOutOfPublicMapsQuota: 'user/isOutOfPublicMapsQuota',
      privateMapsQuota: 'user/privateMapsQuota',
      privateMapsCount: 'user/privateMapsCount',
      isOutOfPrivateMapsQuota: 'user/isOutOfPrivateMapsQuota',
      canCreateMaps: 'user/canCreateMaps'
    }),
    areAllMapsSelected () {
      return Object.keys(this.maps).length === this.selectedMaps.length;
    },
    initialState () {
      return this.isFirstTimeViewingDashboard &&
        !this.hasSharedMaps &&
        !this.isFetchingMaps &&
        this.hasFilterApplied('mine') &&
        this.totalUserEntries <= 0;
    },
    emptyState () {
      return ((!this.isFirstTimeViewingDashboard || this.hasSharedMaps) || this.isFirstTimeViewerAfterAction) &&
        !this.isFetchingMaps &&
        !this.currentEntriesCount;
    },
    emptyStateText () {
      const route = this.$router.resolve({name: 'maps', params: { filter: 'shared' }});

      return this.hasSharedMaps
        ? this.$t('MapsPage.emptyCase.onlyShared', { path: route.href })
        : this.$t('MapsPage.emptyCase.default', { path: route.href });
    },
    isFirstTimeViewerAfterAction () {
      // First time viewing dashboard but user has performed any action such as drag and dropping a dataset (no page refreshing)
      return this.isFirstTimeViewingDashboard && this.currentEntriesCount <= 0 && !this.hasFilterApplied('mine');
    },
    hasSharedMaps () {
      return this.totalShared > 0;
    },
    isSomeMapSelected () {
      return this.selectedMaps.length > 0;
    },
    shouldShowLimitsWarning () {
      return this.isOutOfPublicMapsQuota || this.isOutOfPrivateMapsQuota;
    },
    shouldShowViewSwitcher () {
      return this.canChangeViewMode && !this.initialState && !this.emptyState && !this.selectedMaps.length;
    },
    shouldShowListHeader () {
      return this.isCondensed && !this.emptyState && !this.initialState;
    },
    isNotificationVisible () {
      return this.$store.getters['user/isNotificationVisible'];
    }
  },
  methods: {
    fetchMaps () {
      this.$store.dispatch('maps/fetch');
    },
    applyFilter (filter) {
      this.$emit('applyFilter', filter);
    },
    applyOrder (orderParams) {
      this.$emit('applyOrder', orderParams);
    },
    toggleSelected ({ map, isSelected, event }) {
      if (this.selectedMaps.length && event.shiftKey) {
        this.doShiftClick(map);
        return;
      }

      if (isSelected) {
        this.lastCheckedItem = map;
        this.selectedMaps.push(map);
        return;
      }

      this.selectedMaps = this.selectedMaps.filter(selectedMap => selectedMap.id !== map.id);
    },
    doShiftClick (map) {
      const mapsArray = [...Object.values(this.maps)];
      this.selectedMaps = shiftClick(mapsArray, this.selectedMaps, map, this.lastCheckedItem || map);
    },
    selectAll () {
      this.selectedMaps = [...Object.values(this.$store.state.maps.list)];
    },
    deselectAll () {
      this.selectedMaps = [];
    },
    isMapSelected (map) {
      return this.selectedMaps.some(selectedMap => selectedMap.id === map.id);
    },
    hasFilterApplied (filter) {
      return this.filterType === filter;
    },
    toggleViewMode () {
      this.isCondensed = !this.isCondensed;
    },
    getHeaderContainer () {
      return this.$refs.headerContainer;
    },
    onContentChanged (type) {
      this.$emit('contentChanged', type);
    }
  },
  watch: {
    isCondensed (isCompactMapView) {
      if (isCompactMapView) {
        localStorage.mapViewMode = 'compact';
      } else {
        localStorage.mapViewMode = 'standard';
      }
    },
    selectedMaps () {
      this.$emit('selectionChange', this.selectedMaps);
    }
  }
};
</script>

<style scoped lang="scss">
@import 'new-dashboard/styles/variables';

.map-element {
  margin-bottom: 36px;
}

.full-width {
  width: 100%;
}

.grid__head--sticky {
  top: $header__height;
}

.grid__head--sticky.has-user-notification {
  top: $header__height + $notification-warning__height;
}

.pagination-element {
  margin-top: 28px;
}

.empty-state {
  margin: 20vh 0 8vh;
}

.grid-column {
  flex-direction: column;
}

.card-condensed {
  width: 100%;
  border-bottom: 1px solid #EBEEF5;

  &:last-child {
    border-bottom: 0;
  }
}

.mapcard-view-mode {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 38px;
  height: 36px;
  margin-left: 24px;
  padding: 9px;
  cursor: pointer;

  &:hover,
  &:focus {
    background-color: $softblue;
  }

  &:active {
    background-color: $primary-color;

    .svgicon {
      fill: $white;
    }
  }
}

.warning {
  white-space: nowrap;
}
</style>
