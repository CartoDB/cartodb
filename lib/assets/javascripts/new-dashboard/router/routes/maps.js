// Hooks
import { mapsBeforeEnter, externalMapsBeforeEnter } from 'new-dashboard/router/hooks/check-navigation';

// Lazy Pages
const Maps = () => import('new-dashboard/pages/Maps/Maps');
const Carto = () => import('new-dashboard/pages/Maps/Carto');
const External = () => import('new-dashboard/pages/Maps/Carto');

const routes = [
  {
    path: '/maps/:filter?',
    component: Maps,
    meta: {
      title: () => 'Maps | CARTO'
    },
    children: [
      {
        path: '',
        name: 'maps',
        component: Carto,
        meta: {
          title: () => 'Maps | CARTO'
        },
        beforeEnter: mapsBeforeEnter
      },
      {
        path: 'external',
        name: 'external',
        component: External,
        meta: {
          title: () => 'Kepler.gl Maps | CARTO'
        },
        beforeEnter: externalMapsBeforeEnter
      }
    ]
  },
  {
    path: '/maps/external/:filter',
    name: 'external_filtered',
    component: Maps,
    meta: {
      title: () => 'Kepler.gl  Maps | CARTO'
    },
    children: [
      {
        path: '',
        name: 'external_filtered',
        component: External,
        meta: {
          title: () => 'Kepler.gl  Maps | CARTO'
        },
        beforeEnter: externalMapsBeforeEnter
      }
    ]
  }
];

export default routes;
