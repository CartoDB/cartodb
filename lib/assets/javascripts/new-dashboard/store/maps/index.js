import CartoNode from 'carto-node';
import toObject from 'new-dashboard/utils/to-object';
import featuredFavoritedMaps from 'new-dashboard/store/maps/featured-favorited-maps';
import Filters, { defaultParams as filtersDefaultParams } from 'new-dashboard/core/filters';

const client = new CartoNode.AuthenticatedClient();

const DEFAULT_VALUES = {
  filter: 'mine',
  order: 'favorited,updated_at',
  orderDirection: 'desc,desc'
};

const maps = {
  namespaced: true,
  modules: {
    featuredFavoritedMaps
  },
  state: {
    isFetching: false,
    isFiltered: false,
    isErrored: false,
    error: {},
    filterType: DEFAULT_VALUES.filter,
    order: DEFAULT_VALUES.order,
    orderDirection: DEFAULT_VALUES.orderDirection,
    list: {},
    metadata: {},
    page: 1,
    numPages: 1,
    resultsPerPage: 12
  },
  getters: {},
  mutations: {
    setRequestError (state, error) {
      state.isFetching = false;
      state.isErrored = true;
      state.error = error;
    },

    setMaps (state, maps) {
      state.list = toObject(maps.visualizations, 'id');
      state.metadata = {
        total_entries: maps.total_entries,
        total_likes: maps.total_likes,
        total_shared: maps.total_shared,
        total_locked: maps.total_locked,
        total_user_entries: maps.total_user_entries
      };

      state.isFetching = false;
    },

    setFetchingState (state) {
      state.isFetching = true;
      state.isErrored = false;
      state.error = {};
    },

    setFilterType (state, filterType = 'mine') {
      state.filterType = filterType;
    },

    setPagination (state, page = 1) {
      state.page = page;
      state.numPages = Math.ceil(state.metadata.total_entries / filtersDefaultParams.per_page) || 1;
    },

    setMapAttributes (state, {mapId, mapAttributes}) {
      Object.assign(state.list[mapId], mapAttributes);
    },

    setOrder (state, orderOpts) {
      state.order = orderOpts.order || DEFAULT_VALUES.order;
      state.orderDirection = orderOpts.direction || DEFAULT_VALUES.orderDirection;
    },

    setResultsPerPage (state, resultsPerPage = 12) {
      state.resultsPerPage = resultsPerPage;
    },

    updateLike (state, {mapId, liked}) {
      state.list[mapId].liked = liked;
    },

    updateNumberLikes (state, {mapId, likes}) {
      state.list[mapId].likes = likes;
    }
  },
  actions: {
    fetchMaps (context) {
      context.commit('setFetchingState');
      const params = {
        ...Filters[context.state.filterType],
        types: 'derived',
        page: context.state.page,
        order: context.state.order,
        order_direction: context.state.orderDirection,
        per_page: context.state.resultsPerPage
      };

      client.getVisualization('',
        params,
        function (err, _, data) {
          if (err) {
            context.commit('setRequestError', err);
            return;
          }
          context.commit('setMaps', data);
          context.commit('setPagination', context.state.page);
        }
      );
    },
    goToPage (context, page) {
      context.commit('setPagination', page);
      context.dispatch('fetchMaps');
    },
    filterMaps (context, filter) {
      context.commit('setPagination', 1);
      context.commit('setFilterType', filter);
      context.dispatch('fetchMaps');
    },
    orderMaps (context, orderOpts) {
      context.commit('setPagination', 1);
      context.commit('setOrder', orderOpts);
      context.dispatch('fetchMaps');
    },
    resetFilters (context) {
      context.commit('setPagination', 1);
      context.commit('setFilterType', 'mine');
      context.commit('setResultsPerPage', 12);
      context.commit('setOrder', {order: DEFAULT_VALUES.order, orderDirection: DEFAULT_VALUES.orderDirection});
    },
    setPerPage (context, resultsPerPage) {
      context.commit('setResultsPerPage', resultsPerPage);
    },
    updateMap (context, mapOptions) {
      context.commit('setMapAttributes', mapOptions);
    },
    setURLOptions (context, options) {
      context.commit('setPagination', options.page);
      context.commit('setFilterType', options.filter);
      context.commit('setOrder', {direction: options.order_direction, order: options.order});
      context.dispatch('fetchMaps');
    },
    like (context, map) {
      const currentLikeStatus = map.liked;
      context.commit('updateLike', { mapId: map.id, liked: true });
      client.like(map.id,
        function (err, _, data) {
          if (err) {
            context.commit('updateLike', { mapId: map.id, liked: currentLikeStatus });
            return;
          }
          context.commit('updateNumberLikes', { mapId: map.id, likes: data.likes });
        }
      );
    },
    deleteLike (context, map) {
      const currentLikeStatus = map.liked;
      context.commit('updateLike', { mapId: map.id, liked: false });
      client.deleteLike(map.id,
        function (err, _, data) {
          if (err) {
            context.commit('updateLike', { mapId: map.id, liked: currentLikeStatus });
            return;
          }
          context.commit('updateNumberLikes', { mapId: map.id, likes: data.likes });
        }
      );
    }
  }
};

export default maps;
