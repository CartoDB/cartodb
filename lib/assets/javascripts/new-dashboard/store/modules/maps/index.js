import toObject from 'new-dashboard/utils/to-object';
import featuredFavoritedMaps from 'new-dashboard/store/modules/maps/featured-favorited-maps';
import Filters, { defaultParams as filtersDefaultParams } from 'new-dashboard/core/configuration/filters';

import * as VisualizationActions from '../../actions/visualizations';
import * as VisualizationMutations from '../../mutations/visualizations';

const DEFAULT_VALUES = {
  filter: 'mine',
  order: 'favorited,updated_at',
  orderDirection: 'desc,desc'
};

const maps = {
  namespaced: true,
  modules: {
    featuredFavoritedMaps
  },
  state: {
    isFetching: false,
    isFiltered: false,
    isErrored: false,
    error: {},
    filterType: DEFAULT_VALUES.filter,
    order: DEFAULT_VALUES.order,
    orderDirection: DEFAULT_VALUES.orderDirection,
    list: {},
    metadata: {},
    page: 1,
    numPages: 1,
    resultsPerPage: 12
  },
  getters: {},
  mutations: {
    setRequestError (state, error) {
      state.isFetching = false;
      state.isErrored = true;
      state.error = error;
    },

    setVisualizations (state, maps) {
      state.list = toObject(maps.visualizations, 'id');
      state.metadata = {
        total_entries: maps.total_entries,
        total_likes: maps.total_likes,
        total_shared: maps.total_shared,
        total_locked: maps.total_locked,
        total_user_entries: maps.total_user_entries
      };

      state.isFetching = false;
    },

    setFetchingState (state) {
      state.isFetching = true;
      state.isErrored = false;
      state.error = {};
    },

    setFilterType (state, filterType = 'mine') {
      state.filterType = filterType;
    },

    setPagination (state, page = 1) {
      state.page = page;
      state.numPages = Math.ceil(state.metadata.total_entries / filtersDefaultParams.per_page) || 1;
    },

    setOrder (state, orderOpts) {
      state.order = orderOpts.order || DEFAULT_VALUES.order;
      state.orderDirection = orderOpts.direction || DEFAULT_VALUES.orderDirection;
    },

    setResultsPerPage (state, resultsPerPage = 12) {
      state.resultsPerPage = resultsPerPage;
    },

    updateLike: VisualizationMutations.updateLike,
    updateNumberLikes: VisualizationMutations.updateNumberLikes
  },
  actions: {
    fetch (context) {
      const params = {
        ...Filters[context.state.filterType],
        types: 'derived',
        page: context.state.page,
        order: context.state.order,
        order_direction: context.state.orderDirection,
        per_page: context.state.resultsPerPage
      };

      VisualizationActions.fetchVisualizations(context, params);
    },
    goToPage (context, page) {
      context.commit('setPagination', page);
      context.dispatch('fetch');
    },
    filterMaps (context, filter) {
      context.commit('setPagination', 1);
      context.commit('setFilterType', filter);
      context.dispatch('fetch');
    },
    orderMaps (context, orderOpts) {
      context.commit('setPagination', 1);
      context.commit('setOrder', orderOpts);
      context.dispatch('fetch');
    },
    resetFilters (context) {
      context.commit('setPagination', 1);
      context.commit('setFilterType', 'mine');
      context.commit('setResultsPerPage', 12);
      context.commit('setOrder', {order: DEFAULT_VALUES.order, orderDirection: DEFAULT_VALUES.orderDirection});
    },
    setPerPage (context, resultsPerPage) {
      context.commit('setResultsPerPage', resultsPerPage);
    },
    updateVisualization: VisualizationActions.updateVisualization,
    setURLOptions (context, options) {
      context.commit('setPagination', options.page);
      context.commit('setFilterType', options.filter);
      context.commit('setOrder', {direction: options.order_direction, order: options.order});
      context.dispatch('fetch');
    },
    like: VisualizationActions.like,
    deleteLike: VisualizationActions.deleteLike
  }
};

export default maps;
