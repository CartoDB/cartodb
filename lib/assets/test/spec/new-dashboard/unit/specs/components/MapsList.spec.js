import { createLocalVue, shallowMount } from '@vue/test-utils';
import Vuex from 'vuex';
import maps from '../fixtures/visualizations';
import MapsList from 'new-dashboard/components/MapsList';

jest.mock('new-dashboard/components/CreateButton', () => {});

const localVue = createLocalVue();
localVue.use(Vuex);

describe('MapsList.vue', () => {
  it('should render properly with condensed card and without change view mode available', () => {
    const store = createStore();
    const propsData = {
      hasBulkActions: true,
      canHoverCard: true,
      canChangeViewMode: false,
      isCondensedDefault: true,
      showViewAll: true
    };
    const mapsList = createMapsList(store, propsData);

    expect(mapsList).toMatchSnapshot();
  });

  it('should render properly with simple card and with change view mode available', () => {
    const store = createStore();
    const propsData = {
      hasBulkActions: true,
      canHoverCard: true,
      canChangeViewMode: true,
      isCondensedDefault: false,
      showViewAll: true
    };
    const mapsList = createMapsList(store, propsData);

    expect(mapsList).toMatchSnapshot();
  });
  it('should render initial state', () => {
    const store = createStore({
      maps: {
        isFetchingMaps: false,
        list: [],
        metadata: {
          total_user_entries: 0
        },
        filterType: 'mine'
      },
      config: {
        isFirstTimeViewingDashboard: false
      }
    });
    const mapsList = createMapsList(store);

    expect(mapsList).toMatchSnapshot();
  });

  it('should render empty state', () => {
    const store = createStore({
      maps: {
        isFetchingMaps: false,
        list: [],
        metadata: {
          total_entries: 0,
          total_user_entries: 2
        },
        filterType: 'mine'
      },
      config: {
        isFirstTimeViewingDashboard: false
      }
    });
    const mapsList = createMapsList(store);

    expect(mapsList).toMatchSnapshot();
  });

  describe('Methods', () => {
    it('applyOrder: should dispatch maps/order with order options', () => {
      const store = createStore();
      const mapsList = createMapsList(store);

      mapsList.vm.applyOrder({ order: 'updated_at', direction: 'asc' });

      expect(mapsList.emitted('applyOrder').length).toBe(1);
    });

    it('applyFilter: should dispatch maps/filter with filter type', () => {
      const store = createStore();
      const mapsList = createMapsList(store);

      mapsList.vm.applyFilter('locked');

      expect(mapsList.emitted('applyFilter').length).toBe(1);
    });

    it('fetchMaps: should dispatch maps/fetch', () => {
      const store = createStore();
      spyOn(store, 'dispatch');
      const mapsList = createMapsList(store);

      mapsList.vm.fetchMaps();

      expect(store.dispatch).toHaveBeenCalledWith('maps/fetch');
    });

    it('hasFilterApplied: should return true if applied filter is equal to the one passed', () => {
      const store = createStore();
      const mapsList = createMapsList(store);

      expect(mapsList.vm.hasFilterApplied('mine')).toBe(true);
    });

    it('toggleSelected: adds the map passed to selected maps list', () => {
      const store = createStore();
      const propsData = {
        canHoverCard: true
      };
      const mapsList = createMapsList(store, propsData);

      const map = maps.visualizations[0];
      const event = {
        shiftKey: shiftKey => false
      };
      const isSelected = true;
      mapsList.vm.toggleSelected({ map, isSelected, event });

      expect(mapsList.vm.selectedMaps[0]).toEqual(map);
      expect(mapsList.vm.lastCheckedItem).toEqual(map);
    });

    it('selectAll: marks all maps as selected', () => {
      const store = createStore();
      const mapsList = createMapsList(store);
      mapsList.setProps({
        canHoverCard: true
      });
      mapsList.setProps({
        selectedMaps: []
      });
      mapsList.vm.selectAll();

      expect(mapsList.vm.selectedMaps).toEqual(maps.visualizations);
    });

    it('isMapSelected: returns true if dataset is selected', () => {
      const map = maps.visualizations[0];
      const store = createStore();
      const mapsList = createMapsList(store);
      mapsList.setData({
        selectedMaps: [map]
      });
      expect(mapsList.vm.isMapSelected(map)).toBe(true);
    });

    it('onContentChanged: should emit a contentChanged event when DatasetCard emits it', () => {
      const store = createStore();
      const mapsList = createMapsList(store);

      mapsList.vm.onContentChanged();

      expect(mapsList.emitted('contentChanged')).toBeTruthy();
    });
  });
});

function createStore (customStoreData) {
  const storeData = {
    maps: {
      isFetchingMaps: false,
      list: maps.visualizations,
      metadata: {
        total_user_entries: 2,
        total_entries: 2
      },
      filterType: 'mine',
      order: 'updated_at',
      orderDirection: 'desc'
    },
    config: {
      isFirstTimeViewingDashboard: false
    }
  };

  const storeInstance = new Vuex.Store({
    state: customStoreData || storeData
  });

  return storeInstance;
}

function createMapsList (store, propsData) {
  return shallowMount(MapsList, {
    mocks: {
      $t: key => key,
      $router: {
        resolve: href => href
      }
    },
    store,
    localVue,
    propsData
  });
}
