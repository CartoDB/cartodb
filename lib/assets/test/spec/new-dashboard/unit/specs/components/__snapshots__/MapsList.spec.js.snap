// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MapsList.vue should render empty state 1`] = `
<div class="container grid">
  <div class="full-width">
    <sectiontitle-stub showactionbutton="true" class="grid-cell"><template></template> <template></template> <template></template> <template></template> <template></template></sectiontitle-stub>
    <!---->
    <!---->
    <!---->
    <!---->
    <emptystate-stub text="MapsPage.emptyCase.default"><img svg-inline="" src="../assets/icons/common/compass.svg"></emptystate-stub>
  </div>
</div>
`;

exports[`MapsList.vue should render initial state 1`] = `
<div class="container grid">
  <div class="full-width">
    <sectiontitle-stub showactionbutton="true" class="grid-cell"><template></template> <template></template> <template></template> <template></template> <template></template></sectiontitle-stub>
    <!---->
    <!---->
    <!---->
    <!---->
    <emptystate-stub text="MapsPage.emptyCase.default"><img svg-inline="" src="../assets/icons/common/compass.svg"></emptystate-stub>
  </div>
</div>
`;

exports[`MapsList.vue should render properly with condensed card and without change view mode available 1`] = `
<div class="container grid" showviewall="true">
  <div class="full-width">
    <sectiontitle-stub showactionbutton="true" class="grid-cell"><template></template> <template></template> <template></template> <template></template> <template></template></sectiontitle-stub>
    <!---->
    <div class="grid-cell grid-cell--noMargin grid-cell--col12 grid__head--sticky">
      <condensedmapheader-stub order="updated_at" orderdirection="desc" issortable="true" showviews="true"></condensedmapheader-stub>
    </div>
    <!---->
    <ul class="grid grid-column grid-cell">
      <li class="card-condensed">
        <mapcard-stub visualization="[object Object]" storeactiontype="maps" canhover="true" condensed="true"></mapcard-stub>
      </li>
      <li class="card-condensed">
        <mapcard-stub visualization="[object Object]" storeactiontype="maps" canhover="true" condensed="true"></mapcard-stub>
      </li>
    </ul>
    <!---->
  </div>
</div>
`;

exports[`MapsList.vue should render properly with simple card and with change view mode available 1`] = `
<div class="container grid" showviewall="true">
  <div class="full-width">
    <sectiontitle-stub showactionbutton="true" class="grid-cell"><template></template> <template></template> <template></template> <template></template> <template></template></sectiontitle-stub>
    <!---->
    <!---->
    <!---->
    <ul class="grid">
      <li class="grid-cell grid-cell--col4 grid-cell--col6--tablet grid-cell--col12--mobile map-element">
        <mapcard-stub visualization="[object Object]" storeactiontype="maps" canhover="true"></mapcard-stub>
      </li>
      <li class="grid-cell grid-cell--col4 grid-cell--col6--tablet grid-cell--col12--mobile map-element">
        <mapcard-stub visualization="[object Object]" storeactiontype="maps" canhover="true"></mapcard-stub>
      </li>
    </ul>
    <!---->
  </div>
</div>
`;
